ext/opcache/Optimizer/zend_inference.c

```diff
@@ -2111,11 +2111,7 @@ static uint32_t assign_dim_result_type(
 				tmp |= MAY_BE_ARRAY_KEY_LONG;
 			}
 			if (dim_type & MAY_BE_STRING) {
-				tmp |= MAY_BE_ARRAY_KEY_STRING;
-				if (dim_op_type != IS_CONST) {
-					// FIXME: numeric string
-					tmp |= MAY_BE_ARRAY_KEY_LONG;
-				}
+				tmp |= (MAY_BE_ARRAY_KEY_LONG | MAY_BE_ARRAY_KEY_STRING);
```

Comment:
These need to be changed to replace the `dim_op_type != IS_CONST` check with a ZEND_HANDLE_NUMERIC for CONST operands.

Reply:
It would make sense to extract the whole key type checking into a separate function, as this code is repeated a couple of times.

You'll need to pass in the znode_op* for the dim next to dim_op_type.

Reply:
So if `dim_op_type == IS_CONST`, we should do a `ZEND_HANDLE_NUMERIC ` to add `MAY_BE_ARRAY_KEY_LONG`? 

I'm really just following your lead here but I'll try to make sense of this. If the dimension operand is a constant (I'm guessing a literal string in our case), we should check if it's numeric.. but why? What is `tmp` here and who are we signalling that `tmp` `MAY_BE_ARRAY_KEY_LONG`?

Also, what's a `znode_op`?

Reply:
If dim_op_type != IS_CONST: You always add MAY_BE_ARRAY_KEY_LONG.
If dim_op_type == IS_CONST: You do a ZEND_HANDLE_NUMERIC and only add MAY_BE_KEY_LONG if it's numeric.

To avoid the need to pass in the full kitchen sink of parameters to this function, you might want to pass `opline->op2_type == IS_CONST ? CRT_CONSTANT_EX(op_array, opline, opline->op2, ssa->rt_constants) : NULL` from the caller, rather than just `opline->op2` (which is the `znode_op` I was referring to).